### PR TITLE
Write mesh file for mpm input

### PR DESCRIPTION
**Pull request by ezrayst**
- [x] Write mesh file for mpm input
- [ ] Create tests to validate cases with multiple entities for nodes, curves, faces and volumes

### CODE REVIEW

File: include/mesh.tcc [hunk 1]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+//! \brief Write node coordinates and element id
+//! \details Generate mesh file for mpm code
+//! \tparam Tdim dimension
+//! \tparam Tvertices Number of vertices in an element
+template <unsigned Tdim, unsigned Tvertices>
+void Mesh<Tdim, Tvertices>::write_mpm_mesh(
+    const boost::filesystem::path& mesh_mpm_filename) {
+
+  const auto filename = mesh_mpm_filename.string();
+  std::cout << "mesh file for mpm will be stored in: " << filename << '\n';
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

Comment (kks32) [reply include/mesh.tcc]: This is going to be in the `results/` directory no need to print out.

File: include/mesh.tcc [hunk 2]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+  //! Output vertices file
+  std::fstream mesh_mpm_file;
+  mesh_mpm_file.open(filename, std::ios::out);
+
+  if (mesh_mpm_file.is_open()) {
+    //! Write the total number of vertices
+    mesh_mpm_file << vertices_.size() << '\t' << elements_.size() << '\n';
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
    mesh_mpm_file << vertices_.size() << "\t" << elements_.size() << "\n";
```

File: include/mesh.tcc [hunk 3]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+    //! Write the coordinates of the vertices
+    //! [X] [Y] [Z]
+    //! Note that for 2D, there are no z values
```
Comment (kks32) [inline include/mesh.tcc:331]: There are no Z coordinates in our 2D case as an input to the mesh.

File: include/mesh.tcc [hunk 4]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+    //! Iterate over vertices to get coordinates
+    for (std::map<unsigned, Eigen::VectorXd>::iterator it = vertices_.begin();
```
Comment (kks32) [inline include/mesh.tcc]: Use `const auto`

Comment (kks32) [reply include/mesh.tcc]: `it` should be `itr`

File: include/mesh.tcc [hunk 5]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+         it != vertices_.end(); ++it) {
+      Eigen::VectorXd vertice = it->second;
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
      Eigen::VectorXd vertex = it->second;
```

File: include/mesh.tcc [hunk 6]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+      mesh_mpm_file << vertice[0] << '\t' << vertice[1] << '\t' << vertice[2]
```
Comment (kks32) [inline include/mesh.tcc]: Can you write this in a for-loop so it works for both 2D and 3D cases?

File: include/mesh.tcc [hunk 7]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+                    << '\n';
+    }
+
+    //! Write the element id
+    //! It follows GMSH4 convention
+    const unsigned nodes_in_element = 8;
```
Comment (kks32) [inline include/mesh.tcc]: This will only for a very special case of 3D 8-noded element?

File: include/mesh.tcc [hunk 8]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+    for (const auto& element : elements_) {
+      Eigen::VectorXd vertices = element->vertices();
+      for (unsigned i = 0; i < nodes_in_element; ++i) {
+        mesh_mpm_file << vertices[i] - 1 << '\t';
```
Comment (kks32) [inline include/mesh.tcc]: Is it because of numbering inside GMSH? Could you add a comment to justify that/

File: include/mesh.tcc [hunk 9]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+      mesh_mpm_file << vertice[0] << '\t' << vertice[1] << '\t' << vertice[2]
+                    << '\n';
+    }
+
+    //! Write the element id
+    //! It follows GMSH4 convention
+    const unsigned nodes_in_element = 8;
+    for (const auto& element : elements_) {
+      Eigen::VectorXd vertices = element->vertices();
+      for (unsigned i = 0; i < nodes_in_element; ++i) {
```
Comment (kks32) [inline include/mesh.tcc]: Can we not just iterate over `vertices.size()`?

File: include/mesh.tcc [hunk 10]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+        mesh_mpm_file << vertices[i] - 1 << '\t';
+      }
+      mesh_mpm_file << '\n';
+    }
+
+    mesh_mpm_file.close();
+  }
+  std::cout << "Wrote mesh file\n";
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

File: include/mesh.tcc [hunk 11]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+    //! It follows GMSH4 convention
+    const unsigned nodes_in_element = 8;
+    for (const auto& element : elements_) {
+      Eigen::VectorXd vertices = element->vertices();
+      for (unsigned i = 0; i < nodes_in_element; ++i) {
+        mesh_mpm_file << vertices[i] - 1 << '\t';
+      }
+      mesh_mpm_file << '\n';
+    }
+
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

Comment (kks32) [reply include/mesh.tcc]: Remove unused space

File: include/mesh.tcc [hunk 12]
```diff
@@ -307,3 +307,50 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+    //! Write the element id
+    //! It follows GMSH4 convention
+    const unsigned nodes_in_element = 8;
+    for (const auto& element : elements_) {
+      Eigen::VectorXd vertices = element->vertices();
+      for (unsigned i = 0; i < nodes_in_element; ++i) {
+        mesh_mpm_file << vertices[i] - 1 << '\t';
+      }
+      mesh_mpm_file << '\n';
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
      mesh_mpm_file << "\n";
```

File: include/mesh.tcc (hunk 8 above)
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
        mesh_mpm_file << (vertices[i] - 1) << "\t";
```

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by thiagordonho — APPROVED** [review]

File: include/mesh.tcc (hunk 9 above)
Comment (ezrayst) [reply include/mesh.tcc]: Yes. It's better I think.

File: include/mesh.tcc (hunk 3 above)
Comment (ezrayst) [reply include/mesh.tcc:331]: Refactored :)

File: include/mesh.tcc (hunk 10 above)
Comment (ezrayst) [reply include/mesh.tcc]: We have that for the rest, like writing coordinates, stresses, volumes and so on. If you don't want this, we should remove everything else too.

File: include/gmsh.tcc [hunk 13]
```diff
@@ -175,27 +175,18 @@ void GMSH<Tdim, Tvertices>::read_elements(std::ifstream& file) {
...
-        std::cout << line << "\n";
-        std::cout << istream.str() << "\n";
-        
         if (line.find('#') == std::string::npos && line != "") {
-
           //! If element type not equals to specified Tvertices, skip element
           if (elementtype == element_type) {
             istream >> elementid;
             //! For every element, get the node number of its vertices
             for (unsigned k = 0; k < elementarray.size(); ++k) {
```
Comment (kks32) [inline include/gmsh.tcc]: Remove `{` and `}`

File: include/mesh.h [hunk 14]
```diff
@@ -60,6 +60,9 @@ class Mesh {
   //! Write .vtk files for mesh
   void write_vtk_mesh(const boost::filesystem::path& mesh_vtk_filename);
 
+  //! Write mesh file for mpm input
+  void write_mpm_mesh(const boost::filesystem::path& mesh_vtk_filename);
```
Comment (kks32) [inline include/mesh.h]:
```suggestion
  void write_mpm_mesh(const boost::filesystem::path& mesh_filename);
```

File: include/mesh.tcc [hunk 15]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
   std::cout << "Wrote mesh vtk file\n";
 }
+
+//! \brief Write node coordinates and element id
+//! \details Generate mesh file for mpm code
+//! \tparam Tdim dimension
+//! \tparam Tvertices Number of vertices in an element
+template <unsigned Tdim, unsigned Tvertices>
+void Mesh<Tdim, Tvertices>::write_mpm_mesh(
+    const boost::filesystem::path& mesh_mpm_filename) {
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
    const boost::filesystem::path& mesh_filename) {
```

File: include/mesh.tcc [hunk 16]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+  const auto filename = mesh_mpm_filename.string();
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
  const auto filename = mesh_filename.string();
```

File: include/mesh.tcc [hunk 17]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+  if (mesh_mpm_file.is_open()) {
+    //! Write the total number of vertices
+    mesh_mpm_file << vertices_.size() << "\t" << elements_.size() << "\n";
+
+    //! Write the coordinates of the vertices
+    //! [X] [Y] [Z]
+    //! Note that for 2D, there are no z values
+    //! Iterate over vertices to get coordinates
+    for (auto itr = vertices_.begin(); itr != vertices_.end(); ++itr) {
```
Comment (kks32) [inline include/mesh.tcc]: Could you use range based for loop?

File: include/mesh.tcc [hunk 18]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+      Eigen::VectorXd coordinates = itr->second;
+      for (unsigned i = 0; i < coordinates.size(); ++i) {
+        if (i == (coordinates.size() - 1))
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

File: include/mesh.tcc [hunk 19]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+
+    //! Write the coordinates of the vertices
+    //! [X] [Y] [Z]
+    //! Note that for 2D, there are no z values
+    //! Iterate over vertices to get coordinates
+    for (auto itr = vertices_.begin(); itr != vertices_.end(); ++itr) {
+      Eigen::VectorXd coordinates = itr->second;
+      for (unsigned i = 0; i < coordinates.size(); ++i) {
+        if (i == (coordinates.size() - 1)) 
+          mesh_mpm_file << coordinates[i];
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

File: include/mesh.tcc [hunk 20]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+        else
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
```

File: include/mesh.tcc [hunk 21]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+          mesh_mpm_file << coordinates[i] << '\t';
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
          mesh_mpm_file << coordinates[i] << "\t";
```

File: include/mesh.tcc [hunk 22]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+      }
+      mesh_mpm_file << "\n";
+    }
+
+    //! Write the element id following GMSH4 convention
+    for (const auto& element : elements_) {
+      Eigen::VectorXd vertices = element->vertices();
+      for (unsigned i = 0; i < vertices.size(); ++i) {
```
Comment (kks32) [inline include/mesh.tcc:343]: Remove `{` and `}` in for loop

File: include/mesh.tcc [hunk 23]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+        mesh_mpm_file << (vertices[i] - 1) << '\t';
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
        //  Vertices are numbered from 0 in MPM mesh, while GMSH are numbered from 1
        mesh_mpm_file << (vertices[i] - 1) << "\t";
```

File: include/mesh.tcc [hunk 24]
```diff
@@ -307,3 +307,49 @@ void Mesh<Tdim, Tvertices>::write_vtk_mesh(
...
+      }
+      mesh_mpm_file << "\n";
```
Comment (kks32) [inline include/mesh.tcc]:
```suggestion
      mesh_mpm_file << std::endl;
```

File: include/mesh.tcc (hunk 10 above)
Comment (kks32) [reply include/mesh.tcc]: We can remove the rest in a different PR

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by thiagordonho — APPROVED** [review]

**Review by thiagordonho — APPROVED** [review]

File: include/gmsh.tcc (hunk 13 above)
Comment (ezrayst) [reply include/gmsh.tcc]: Done

File: include/mesh.tcc (hunk 17 above)
Comment (ezrayst) [reply include/mesh.tcc]: Done

File: include/mesh.tcc (hunk 22 above)
Comment (ezrayst) [reply include/mesh.tcc:343]: Done

**Review by kks32 — CHANGES_REQUESTED** [review]
Could you please fix the failing unit test?

**Pull request merged**